### PR TITLE
[ROCm] HipBLASLt row major layout support.

### PR DESCRIPTION
**Pull request by wenchenvincent**
hipBlasLt currently only supports column major. This PR enables using hipBLASLt for matrices of row major layout by swapping the operands and manipulating the transpose attributes.

### CODE REVIEW

**Comment by wenchenvincent** [comment]
@akuegel Could you kindly review this PR?

File: xla/service/gpu/matmul_utils.cc [hunk 1]
```diff
@@ -902,6 +902,17 @@ StatusOr<se::gpu::BlasLt::Epilogue> AsBlasLtEpilogue(
       GetBlasComputationType(lhs_layout.dtype, output_layout.dtype,
                              config.compute_precision));
 
+  #if TENSORFLOW_USE_ROCM
+	if(lhs_layout.order == MatrixLayout::Order::kRowMajor) {
```
Comment (sergachev) [inline xla/service/gpu/matmul_utils.cc:906]: A space is missing between if and (.

File: xla/service/gpu/matmul_utils.cc [hunk 2]
```diff
@@ -902,6 +902,17 @@ StatusOr<se::gpu::BlasLt::Epilogue> AsBlasLtEpilogue(
...
+	  trans_a = se::blas::Transpose::kTranspose;
+	  lhs_layout.Transpose();
+	}
+  	if(rhs_layout.order == MatrixLayout::Order::kRowMajor) {
```
Comment (sergachev) [inline xla/service/gpu/matmul_utils.cc:910]: Same here.

**Comment by sergachev** [comment]
A test would be good, otherwise looks fine.

File: xla/service/gpu/matmul_utils.cc (hunk 1 above)
Comment (wenchenvincent) [reply xla/service/gpu/matmul_utils.cc]: Fixed the formatting.

**Comment by wenchenvincent** [comment]
> A test would be good, otherwise looks fine.
We can test the change with //xla/tests:matrix_ops_simple_test_gpu

`bazel --bazelrc=/usertools/rocm.bazelrc test --config=rocm --cache_test_results=no --test_env=TF_TESTS_PER_GPU=1 --test_env=TF_GPU_COUNT=1 --test_env=HIPBLASLT_LOG_LEVEL=5 --test_env=ROCBLAS_LAYER=2 --test_env=TF_CPP_VMODULE="matmul_utils=5,cublas_lt_matmul=5,hip_blas_lt=5"  --run_under=//tools/ci_build/gpu_build:parallel_gpu_execute -- //xla/tests:matrix_ops_simple_test_gpu`

Unfortunately, we don't have CI setup for ROCm at OpenXLA upstream at this moment.

**Comment by sergachev** [comment]
I understand that CI is missing but how does the change affect matrix_ops_simple_test? Do you mean that currently some of the cases in it are failing and this change fixes some of the failures?

**Comment by wenchenvincent** [comment]
> I understand that CI is missing but how does the change affect matrix_ops_simple_test? Do you mean that currently some of the cases in it are failing and this change fixes some of the failures?

Currently, the matrix_ops_simple_test does not fail but does not really exercise the hipBlasLt path as the code decided that it was not eligible to hipBlasLt if the output matrix was row major layout. So the first commit in the PR enabled using hipBlasLt.

After that was enabled, then two Dot_Add_2x2_2x2F16 subtests in matrix_ops_simple_test would fail. This was due to an issue in hipBlasLt that would require setting the bias pointer before trying to find the algorithms. Otherwise the result would be incorrect. So the second commit in the PR provided a workaround on this issue so that the tests would pass. hipBlasLt will also fix this issue soon.

**Comment by sergachev** [comment]
Our usual approach is to reflect changes in tests, here following your explanation it would be at least to ensure inside the mentioned tests that hipBlasLt is now in use.

**Comment by sergachev** [comment]
Checking test logs is a manual process; if some other change breaks the added support it will go unnoticed. A check should be performed by the test itself, it should fail and prevent breaking changes. Here are some examples:

https://github.com/openxla/xla/blob/6c3ebc1226d1a5e228a7e00c0091937697b1c91e/xla/service/gpu/tests/gemm_rewrite_test.cc#L72

https://github.com/openxla/xla/blob/6c3ebc1226d1a5e228a7e00c0091937697b1c91e/xla/service/gpu/tests/gemm_rewrite_test.cc#L4045

https://github.com/openxla/xla/blob/130c88c2e55e428b7c060f0ef6a11b0e3242a4f4/xla/service/gpu/tests/reduction_vectorization_sm_all.hlo#L4

**Comment by wenchenvincent** [comment]
> Checking test logs is a manual process; if some other change breaks the added support it will go unnoticed. A check should be performed by the test itself, it should fail and prevent breaking changes. Here are some examples:
> 
> https://github.com/openxla/xla/blob/6c3ebc1226d1a5e228a7e00c0091937697b1c91e/xla/service/gpu/tests/gemm_rewrite_test.cc#L72
> 
> https://github.com/openxla/xla/blob/6c3ebc1226d1a5e228a7e00c0091937697b1c91e/xla/service/gpu/tests/gemm_rewrite_test.cc#L4045
> 
> https://github.com/openxla/xla/blob/130c88c2e55e428b7c060f0ef6a11b0e3242a4f4/xla/service/gpu/tests/reduction_vectorization_sm_all.hlo#L4

Thanks for the clarification. I see your point. Let me see if we can add checks like that.

**Comment by wenchenvincent** [comment]
@sergachev I looked into the tests. It seems that the CheckCustomCallTarget test in [xla/xla/service/gpu/tests/gemm_rewrite_test.cc](https://github.com/openxla/xla/blob/6c3ebc1226d1a5e228a7e00c0091937697b1c91e/xla/service/gpu/tests/gemm_rewrite_test.cc) is the test to check cuBlasLt/hipBlasLt is used when xla_gpu_enable_cublaslt=true. Somehow, //xla//service/gpu/tests/gemm_rewrite_test is disabled on ROCm three years and was not run ever since. This is why there was a coverage hole regarding checking whether hipblasLt was run before this PR.

I just re-enabled //xla//service/gpu/tests/gemm_rewrite_test. There were some test failures as there were some gaps in hipBlasLt support so I disabled those tests on ROCm. We will fix those test failures and reinstate them later.

**Comment by sergachev** [comment]
Thank you for looking into tests!

**Review by sergachev — APPROVED** [review]

**Comment by i-chaochen** [comment]
Hi @ddunl wonder how's situation in this PR, is it still requiring another reviewer to approval?

**Review by ddunl — APPROVED** [review]

**Pull request closed without merging**